### PR TITLE
[DO NOT MERGE] Epoch manager draft

### PR DESCRIPTION
**Pull request by pahor167**
Placeholder PR to have the CI

### CODE REVIEW

File: packages/protocol/contracts/common/GoldToken.sol [hunk 1]
```diff
@@ -153,7 +153,7 @@ contract GoldToken is
   function transferFrom(address from, address to, uint256 value) external returns (bool) {
     require(to != address(0), "transfer attempted to reserved address 0x0");
     require(
```
Comment (mcortesi) [inline packages/protocol/contracts/common/GoldToken.sol]: why is this needed? I undersatnd that seems silly, but seems weird that we modify contarct to make sure no one does that

File: packages/protocol/contracts/common/GoldToken.sol [hunk 2]
```diff
@@ -295,7 +296,7 @@ contract GoldToken is
    */
   function _transfer(address to, uint256 value) internal returns (bool) {
     require(
-      to != registry.getAddressForOrDie(CELO_DISTRIBUTION_SCHEDULE_ID),
+      to != registry.getAddressForOrDie(CELO_DISTRIBUTION_SCHEDULE_REGISTRY_ID),
```
Comment (mcortesi) [inline packages/protocol/contracts/common/GoldToken.sol]: same

File: packages/protocol/contracts/common/UsingRegistryV2.sol [hunk 3]
```diff
@@ -178,6 +184,15 @@ contract UsingRegistryV2 {
 
   function getCeloDistributionSchedule() internal view returns (ICeloDistributionSchedule) {
     return
-      ICeloDistributionSchedule(registryContract.getAddressForOrDie(CELO_DISTRIBUTION_SCHEDULE_ID));
+      ICeloDistributionSchedule(registryContract.getAddressForOrDie(CELO_DISTRIBUTION_SCHEDULE_REGISTRY_ID));
+  }
+
+  function getEpochRewards() internal view returns (IEpochRewards) {
```
Comment (mcortesi) [inline packages/protocol/contracts/common/UsingRegistryV2.sol:189]: what is the purpose of UsingRegistryV2 (vs V1)?

heard from Volpe that this was a bad pattern and would better to turn this intro a library, does it make sense to do so for any new contracts?

File: packages/protocol/contracts/governance/interfaces/IEpochRewards.sol [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.5.13 <0.9.0;
+
+interface IEpochRewards {
```
Comment (mcortesi) [inline packages/protocol/contracts/governance/interfaces/IEpochRewards.sol:4]: I suggest we only add here methods that WILL be used on the new system

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol [hunk 5]
```diff
@@ -0,0 +1,278 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.8.0 <0.8.20;
+
+// Note: This is not an exact copy of UsingPrecompiles in the contract's folder, but in solidity 0.8
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "../../contracts/common/interfaces/ICeloVersionedContract.sol";
+import "../common/IsL2Check.sol";
+
+contract UsingPrecompiles is IsL2Check {
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/UsingPrecompiles.sol:10]: found weird that we are adding this class since we don't have precompiles anymore, only the token duality one

File: packages/protocol/contracts-0.8/common/interfaces/IScoreManager.sol [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.8.7 <0.8.20;
+
+interface IScoreManager {
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/interfaces/IScoreManager.sol]: I'd remove uptimes. We're not using them now. So doesn't make sense to have them here.

I also suggest we have a query interface, and we have a write interface.

That is.

ANY score manager would have methods:

 - getValidatorScore(validator) uint256
 - getGroupScore(group) uint256

But depending on the algorithm to compute scores we don't know if ALL score manager implemetnation would have:

 - setValidatorScore
 - setGroupScore


So, i'd say we have a `IScoreReader` with the query methods.

And then a `SimpleScoreManager` that implements it, and have setters for group and validator.

And that's it.

Later if you want a different input strategy you just change the impl. No one else needs to know, since every other part of the system only cares about reading scores

File: packages/protocol/contracts-0.8/common/interfaces/IScoreManager.sol [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+  function setUptimes(address group, uint256[] calldata _uptimes) external;
+  function setValidatorScore(address group, uint256 score) external;
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/interfaces/IScoreManager.sol]: validator, not group

File: packages/protocol/contracts-0.8/common/interfaces/IScoreManager.sol [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+
+  function getUptimes(address group) external view returns (uint256[] memory);
+  function getValidatorScore(address group) external view returns (uint256);
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/interfaces/IScoreManager.sol]: validator, not group

File: packages/protocol/contracts/common/interfaces/IEpochManagerInitializer.sol [hunk 9]
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.5.13 <0.9.0;
+
+interface IEpochManagerInitializer {
```
Comment (mcortesi) [inline packages/protocol/contracts/common/interfaces/IEpochManagerInitializer.sol]: I'm the first to suggest having interfaces.. but this doesn't make much sense to me...

no one dpends on the intializer... no other impls of it, not injected anywhere... so why an interface?

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+  event EpochProcessingStarted(uint256 indexed epochNumber);
+
+  /**
+   * @notice Event emited when epochProcessing has ended.
+   * @param epochNumber The epoch number that is finished being processed.
+   */
+  event EpochProcessingEnded(uint256 indexed epochNumber);
+
+  modifier onlyEpochManagerInitializer() {
+    require(
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:77]: this is an unnecessary coupling...

i don't believe that:

1. the epochManagerInitializer should be on the registry. It's a one time only contract...
2. The epochManager should know about "an intializer"

Instead the epoch Manager has "an owner" or "an initializer" address, that is passed on contruction and is the only one that can call the method

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+  function setEpochDuration(uint256 newEpochDuration) public onlyOwner {
+    epochDuration = newEpochDuration;
+  }
+
+  function isOnEpochProcess() public view returns (bool) {
+    return epochProcessing.status == EpochProcessStatus.Started;
+  }
+
+  function systemAlreadyInitialized() public view returns (bool) {
+    return firstKnownEpoch != 0;
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: is it correct to assume that epoch can start at 0?

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+  }
+
+  // checks if end of epoch has been reached based on timestamp
+  function isReadyToStartEpoch() public view returns (bool) {
+    Epoch memory _currentEpoch = epochs[currentEpochNumber];
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: why so verbose?

can't this be `return block.timestamp > epoch[currentEpochNumber].endTimestamp`?

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+    if (block.timestamp > _currentEpoch.endTimestamp) {
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: this is incorrect...

An epoch ENDS when the the function to transition the epoch is called! 

So, epoch.endTimestamp should NOT be set before that.

The check should be:

```
block.timestamp >= epoch.startTimestamp + epochDuration
```
which is asking if at least minimum time has passed!

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+  // DESIGNDESICION(XXX): we assume that the first epoch on the L2 starts as soon as the system is initialized
+  // to minimize amount of "limbo blocks" the network should stop relatively close to an epoch number (but wigh enough time)
+  // to have time to call the function EpochInitializer.migrateEpochAndValidators()
+  function initializeSystem(
+    uint256 firstEpochNumber,
+    uint256 firstEpochBlock,
+    address[] memory firstElected
+  ) external onlyEpochManagerInitializer {
+    require(!systemAlreadyInitialized(), "Epoch system already initialized");
+    firstKnownEpoch = firstEpochNumber;
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:122]: should check it's not zero

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+    currentEpochNumber = firstEpochNumber;
+
+    Epoch storage _currentEpoch = epochs[currentEpochNumber];
+    _currentEpoch.firstBlock = firstEpochBlock;
+    _currentEpoch.startTimestamp = block.timestamp;
+    _currentEpoch.endTimestamp = block.timestamp + epochDuration;
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: this is incorrect. `endTimestamp` it's when it actually ENDS which you only know when you're closing the epoch

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+
+  // TODO maybe "freezeEpochRewards" "prepareForNextEpoch"
+
+  /// start next epoch process.
+  /// it freezes the epochrewards at the time of execution,
+  /// and starts the distribution of the rewards.
+  function startNextEpochProcess() external nonReentrant {
+    require(isReadyToStartEpoch(), "Epoch is not ready to start");
+    require(!isOnEpochProcess(), "Epoch process is already started");
+    require(elected.length > 0, "Elected length must be greater than 0.");
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: not the place to make this require.. if you need this, then it should be when you're about to SET the elected.

On the initialize and after running elections

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+  /// returns the current epoch number.
+  function getCurrentEpochNumber() external view returns (uint256) {
+    return currentEpochNumber;
+  }
+
+  function getElected() external view returns (address[] memory) {
+    return elected;
+  }
+
+  function getFirstBlockAtEpoch(uint256 epoch) external view returns (uint256) {
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:242]: should check epoch is >= firstKnowEpoch and <= currentEpoch

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+    return epochs[epoch].firstBlock;
+  }
+
+  function getLastBlockAtEpoch(uint256 epoch) external view returns (uint256) {
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:248]: should check epoch is >= firstKnowEpoch and <= currentEpoch

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 19]
```diff
@@ -0,0 +1,317 @@
...
+    elected = firstElected;
+  }
+
+  // TODO maybe "freezeEpochRewards" "prepareForNextEpoch"
+
+  /// start next epoch process.
+  /// it freezes the epochrewards at the time of execution,
+  /// and starts the distribution of the rewards.
+  function startNextEpochProcess() external nonReentrant {
+    require(isReadyToStartEpoch(), "Epoch is not ready to start");
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: should call `isTimeForNextEpoch()` and not this function, which is wrong

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+  function isOnEpochProcess() public view returns (bool) {
+    return epochProcessing.status == EpochProcessStatus.Started;
+  }
+
+  function systemAlreadyInitialized() public view returns (bool) {
+    return firstKnownEpoch != 0;
+  }
+
+  // checks if end of epoch has been reached based on timestamp
+  function isReadyToStartEpoch() public view returns (bool) {
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: remove this function, it's dupplicated and wrong

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+    } else {
+      return false;
+    }
+  }
+
+  function allocateValidatorsRewards() internal {
+    // TODO complete this function
+    uint256 totalRewards = 0;
+    for (uint i = 0; i < elected.length; i++) {
+      uint256 validatorScore = getScoreManager().getValidatorScore(elected[i]);
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: suggest you save the scoreManager on a variable, if not you're making calls to the registry each time

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 22]
```diff
@@ -0,0 +1,317 @@
...
+      uint256 validatorReward = getValidators().computeEpochReward(
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: same

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 23]
```diff
@@ -0,0 +1,317 @@
...
+        elected[i],
+        validatorScore,
+        epochProcessing.perValidatorReward
+      );
+      validatorPendingPayments[elected[i]] += validatorReward;
+      totalRewards += validatorReward;
+    }
+    // Mint all cUSD required for payment and the corresponding CELO
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:306]: This from a "Mento" perspective looks weird.

Some contract is "minting cUSD", and i have no guarantees that it will give the corresponding CELO to the reserve. If i were a Menton protocol developer i wouldn't like this.

On option that i can think of, is to move the logic here (mint CELO, mint cUSD, use the oracle to compute rate, give money to registry) to the CELODistributionSchedule.

So `CeloDistributionSchedule.releaseAscUSD(recipient, amount)`

and CeloDistributionSchedule would have the permission to mint cUSD but it's more believable that it will release CELO as expected. 

Also i don't like the `tranfer` naming for CeloDstirbutionSchedule here

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 24]
```diff
@@ -0,0 +1,317 @@
...
+    require(!systemAlreadyInitialized(), "Epoch system already initialized");
+    firstKnownEpoch = firstEpochNumber;
+    currentEpochNumber = firstEpochNumber;
+
+    Epoch storage _currentEpoch = epochs[currentEpochNumber];
+    _currentEpoch.firstBlock = firstEpochBlock;
+    _currentEpoch.startTimestamp = block.timestamp;
+    _currentEpoch.endTimestamp = block.timestamp + epochDuration;
+
+    elected = firstElected;
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:129]: check lenght >0

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 25]
```diff
@@ -0,0 +1,317 @@
...
+    uint256 firstEpochNumber,
+    uint256 firstEpochBlock,
+    address[] memory firstElected
+  ) external onlyEpochManagerInitializer {
+    require(!systemAlreadyInitialized(), "Epoch system already initialized");
+    firstKnownEpoch = firstEpochNumber;
+    currentEpochNumber = firstEpochNumber;
+
+    Epoch storage _currentEpoch = epochs[currentEpochNumber];
+    _currentEpoch.firstBlock = firstEpochBlock;
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:126]: check it's in the past

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 26]
```diff
@@ -0,0 +1,317 @@
...
+    _currentEpoch.startTimestamp = block.timestamp;
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol:127]: check it is in the past

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 27]
```diff
@@ -0,0 +1,317 @@
...
+    // TODO last block should be the block before and timestamp from previous block
+    epochs[currentEpochNumber].endTimestamp = block.timestamp;
+    epochs[currentEpochNumber].lastBlock = block.number - 1;
+    // start new epoch
+    currentEpochNumber++;
+    epochs[currentEpochNumber].firstBlock = block.number;
+    epochs[currentEpochNumber].startTimestamp = block.timestamp;
+
+    for (uint i = 0; i < elected.length; i++) {
+      (, , address group, , ) = getValidators().getValidator(elected[i]);
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: we might want a 'cheaper' way to getting the group, since here we are getting the whole struct. Also, you want to save the `getValidators()` in variable outside the loop

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 28]
```diff
@@ -0,0 +1,317 @@
...
+      if (!processedGroups[group]) {
+        epochProcessing.toProcessGroups++;
+        processedGroups[group] = true;
+      }
+    }
+
+    require(epochProcessing.toProcessGroups == groups.length, "number of groups does not match");
+
+    for (uint i = 0; i < groups.length; i++) {
+      // checks that group is acutally from elected group
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: typo: Actually

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 29]
```diff
@@ -0,0 +1,317 @@
...
+      require(processedGroups[groups[i]], "group not processed");
+      // by doing this, we avoid processing a group twice
+      delete processedGroups[groups[i]];
+      // TODO what happens to uptime?
+      uint256[] memory uptimes = getScoreManager().getUptimes(groups[i]);
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: Not what we discussed.

should be `scoreManager.getGroupScore(group)`

then modify election contract so that we can call `getGroupEpochRewards(group, groupScore, totalVotersRewards)`

File: packages/protocol/contracts-0.8/common/EpochManager.sol [hunk 30]
```diff
@@ -0,0 +1,317 @@
...
+      uint256 epochRewards = getElection().getGroupEpochRewards(
+        groups[i],
+        epochProcessing.totalRewardsVoter,
+        uptimes
+      );
+      getElection().distributeEpochRewards(groups[i], epochRewards, lessers[i], greaters[i]);
+    }
+    getCeloDistributionSchedule().transfer(
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManager.sol]: I don't like us using `transfer` as the name. Not very indicative of what we are doing.

I Suggest we rename `CeloDistributionSchedule` to `CeloReleaseSchedule` or `CeloUnreleasedTreasure` and name the method `release(beneficiary, amount)`

File: packages/protocol/contracts-0.8/common/EpochManagerInitializer.sol [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+
+    uint256 numberElectedValidators = numberValidatorsInCurrentSet();
+
+    address[] memory electedValidatorAddresses = new address[](numberElectedValidators);
+
+    for (uint256 i = 0; i < numberElectedValidators; i++) {
+      validatorSignerAddressFromCurrentSet(i);
+      electedValidatorAddresses[i] = validatorAddress;
+    }
+    getEpochManager().initializeSystem(currentEpoch, block.number, electedValidatorAddresses);
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/EpochManagerInitializer.sol]: debtable.

instead of `block.number` it should be `getFirstBlockOfEpoch(currentEpoch)`.

For timestamp, you can get an approximate, if you assume every block takes 5 seconds. So check number of blocks from now to first block of epoch, multiply by 5 and substract that to current block.timestamp

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 10 above)
Comment (mcortesi) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:77]: Also if you have an `intialzierAddress, you can set it back to zero after calling this method and with that you make sure you can only be initialized once

File: packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol [hunk 32]
```diff
@@ -41,6 +41,14 @@ contract CeloDistributionSchedule is UsingRegistry, ReentrancyGuard, Initializab
     _;
   }
 
+  modifier onlyEpochManager() {
```
Comment (mcortesi) [inline packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol]: I'm thinking we should completely throw away this contract.

Nothing of the current logic remains. Instead what we have is a CeloUnreleasedTreasury contract that holds unreleased CELO and only allows the Epoch manager to release it with some `releaseXXX()` methods

**Review by mcortesi — COMMENTED** [review]
Left... a "few" comments

File: packages/protocol/contracts/common/GoldToken.sol (hunk 1 above)
Comment (pahor167) [reply packages/protocol/contracts/common/GoldToken.sol]: CeloDistributionSchedule distributes rewards based on its own balance. If the balance is changed from outside by sending Celo to this contract, the distribution schedule will redistribute also this amount. Since we are removing CeloDistributionSchedule logic we will also remove this part.

File: packages/protocol/contracts/common/UsingRegistryV2.sol (hunk 3 above)
Comment (pahor167) [reply packages/protocol/contracts/common/UsingRegistryV2.sol:189]: UsingRegistryV2 is stateless (= it can be added as inherited contract to already existing and deployed core contract without disrupting its state)

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol (hunk 5 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/UsingPrecompiles.sol]: This was a mistake - removed

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 10 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:77]: Agreed - epoch manager initialiser is removed from Registry

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 12 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: Updated to return block.timestamp > _currentEpoch.startTimestamp + epochDuration;

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 15 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: removed

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 16 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: Removed

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 20 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: removed

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 30 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: renamed to release and CeloDistributionSchedule renamed to CeloUnreleasedTreasure

File: packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol (hunk 32 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol]: done

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 18 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:248]: updated with change of < currentEpoch

File: packages/protocol/contracts-0.8/common/interfaces/IScoreManager.sol (hunk 6 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/interfaces/IScoreManager.sol]: Done

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 29 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: done

File: packages/protocol/contracts-0.8/common/UsingPrecompiles.sol (hunk 5 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/UsingPrecompiles.sol:10]: We need this for EpochManagerInitializer so we can calculate epochs correctly - I removed all unnecessary methods from there

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 26 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol:127]: This is a bit problematic since solidity doesn't allow to query block.timestamp from past blocks

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 27 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: new function introduced

File: packages/protocol/contracts-0.8/common/EpochManager.sol (hunk 11 above)
Comment (pahor167) [reply packages/protocol/contracts-0.8/common/EpochManager.sol]: Updated to check 
` return initialized && epochManagerInitializer == address(0);`

File: packages/protocol/contracts/common/interfaces/IEpochManagerInitializer.sol (hunk 9 above)
Comment (pahor167) [reply packages/protocol/contracts/common/interfaces/IEpochManagerInitializer.sol]: Removed

**Comment by martinvol** [comment]
Moved to https://github.com/celo-org/celo-monorepo/pull/11189
